File: [projectx]/[core]/ev-lib/server/sv_rpc.lua
local Resource, Promises, CallIdentifier = GetCurrentResourceName(), {}, 0

RPC = {}

function ClearPromise(callID)
    Citizen.SetTimeout(5000, function()
        Promises[callID] = nil
    end)
end

function ParamPacker(...)
    local params, pack = {...} , {}

    for i = 1, 15, 1 do
        pack[i] = {param = params[i]}
    end

    return pack
end

function ParamUnpacker(params, index)
    local idx = index or 1

    if idx <= #params then
        return params[idx]["param"], ParamUnpacker(params, idx + 1)
    end
end

function UnPacker(params, index)
    local idx = index or 1

    if idx <= 15 then
        return params[idx], UnPacker(params, idx + 1)
    end
end

function TriggerNetEvent(pEvent, pSrc, ...)
    --print("[RPC] Triggering net event", pEvent, pSrc)
    local payload = msgpack.pack({...})

    if payload:len() < 5000 then
        TriggerClientEventInternal(pEvent, pSrc, payload, payload:len())
    else
        TriggerLatentClientEventInternal(pEvent, pSrc, payload, payload:len(), 128000)
    end
end

------------------------------------------------------------------
--                  (Trigger Server Calls)
------------------------------------------------------------------

function RPC.execute(name, src, ...)
    --print("[RPC] Executing", name, src)
    local callID, solved = CallIdentifier, false
    CallIdentifier = CallIdentifier + 1

    Promises[callID] = promise:new()

    TriggerNetEvent("rpc:request:" .. name, src, Resource, callID, ParamPacker(...), true)

    Citizen.SetTimeout(20000, function()
        if not solved then
            Promises[callID]:resolve({nil})
            TriggerEvent("rpc:server:timeout", Resource, name)
        end
    end)

    local response = Citizen.Await(Promises[callID])

    solved = true

    ClearPromise(callID)

    return ParamUnpacker(response)
end

RPC.executeLatent = RPC.execute

RegisterServerEvent("rpc:response")
AddEventHandler("rpc:response", function(origin, callID, response, state)
    if Resource == origin and Promises[callID] then
        --print("[RPC] Response", origin, callID, response, state)
        Promises[callID]:resolve(response)
    end
end)

RegisterServerEvent("rpc:server:timeout")
AddEventHandler("rpc:server:timeout",function(origin, name)
    --print("RPC [Timeout]",origin, name)
end)

RegisterServerEvent("rpc:client:error")
AddEventHandler("rpc:client:error",function(Resource, origin, name, error)
    print("RPC [Error]",Resource, origin, name, error)
end)

------------------------------------------------------------------
--                  (Receive Remote Calls)
------------------------------------------------------------------

function RPC.register(name, func)
    --print("[RPC] Registering", name)
    RegisterNetEvent("rpc:request:" .. name, function(origin, callID, params, packaged)
        --print("[RPC] Request", name, origin, callID, params, packaged)
        local src = source
        local response

        local success, error = pcall(function()
            if packaged then
                response = ParamPacker(func(src, ParamUnpacker(params)))
            else
                response = ParamPacker(func(src, UnPacker(params)))
            end
        end)

        if not success then
            TriggerNetEvent("rpc:client:error", src, Resource, origin, name, error)
        end

        if response == nil then
            response = {}
        end

        TriggerNetEvent("rpc:response", src, origin, callID, response, true)
    end)
end

function RPC.remove(name)
end

File: [projectx]/[core]/ev-lib/server/sv_rpc.js
const Promises = new Map();
const Resource = GetCurrentResourceName();
let CallIdentifier = 0;

function TriggerNetEvent(pEvent, pSrc, ...params) {
    //console.log("[RPC] Triggering net event", pEvent, pSrc);
    const payload = msgpack_pack(params)
    payload.length < 5000
        ? TriggerClientEventInternal(pEvent, pSrc, payload, payload.length)
        : TriggerLatentClientEventInternal(
            pEvent,
            pSrc,
            payload,
            payload.length,
            128000
        )
}

function ParamPacker(...params) {
    const packed = []
    for (let idx = 0; idx < params.length; idx += 1) {
        packed[idx] = { param: params[idx] }
    }
    return packed
}

function ParamUnpacker(params) {
    const unpacked = [];
    params.forEach((p) => unpacked.push(p.param));
    for (let i = 0; i < unpacked.length; i++) {
        if (unpacked[i] == null) {
            unpacked.splice(i, 1);
            i--;
        }
    }
    return unpacked;
}

async function register(name, func) {
    //console.log("[RPC] Registering", name);
    onNet(
        'rpc:request:' + name,
        async function (origin, callID, params, success) {
            //console.log("[RPC] Request", name, origin, callID, success);
            const src = source;
            let response
            success && (params = ParamUnpacker(params))
            try {
                response = await func(src, ...params)
            } catch (error) {
                emit(
                    'rpc:client:error',
                    Resource,
                    origin,
                    name,
                    error.message
                )
            }
            if (typeof response === 'undefined') {
                response = []
            } else {
                success && (response = ParamPacker(response))
            }
            TriggerNetEvent('rpc:response', src, origin, callID, response, success)
        }
    )
}

async function execute(name, src, ...params) {
    //console.log("[RPC] Executing", name, src);
    const callID = CallIdentifier;
    CallIdentifier++;
    
    const pPromise1 = new Promise((resolve, reject) => {
        Promises.set(callID, {
            resolve: resolve,
            reject: reject,
        });
    });
    
    TriggerNetEvent('rpc:request:' + name, src, Resource, callID, params, false);
    
    const pPromise2 = new Promise((resolve) => {
        setTimeout(() => {
            if (Promises.has(callID)) {
                Promises.delete(callID);
            }
            return resolve([]);
        }, 20000)
    });

    return await Promise.race([pPromise1, pPromise2]);
}

function response(origin, callID, response, state) {
    //console.log("[RPC] Response", origin, callID, response, state);
    if (Resource == origin && Promises.has(callID)) {
        state && (response = ParamUnpacker(response));
        Promises.get(callID).resolve(response);
        Promises.delete(callID);
    }
}

const RPC = {
    register: register,
    execute: execute,
}

onNet('rpc:response', response);